src/ModelContextProtocol/Server/McpServer.cs

```diff
@@ -210,21 +185,37 @@ private void SetResourcesHandler(McpServerOptions options)
 
         var listResourcesHandler = resourcesCapability.ListResourcesHandler;
         var listResourceTemplatesHandler = resourcesCapability.ListResourceTemplatesHandler;
+        var readResourceHandler = resourcesCapability.ReadResourceHandler;
+        var resourceCollection = resourcesCapability.ResourceCollection;
 
-        if ((listResourcesHandler is not { } && listResourceTemplatesHandler is not { }) ||
-            resourcesCapability.ReadResourceHandler is not { } readResourceHandler)
+        var originalListResourcesHandler = listResourcesHandler;
+        listResourcesHandler = async (request, cancellationToken) =>
         {
-            throw new McpException("Resources capability was enabled, but ListResources and/or ReadResource handlers were not specified.");
-        }
+            ListResourcesResult result = originalListResourcesHandler is not null ?
+                await originalListResourcesHandler(request, cancellationToken).ConfigureAwait(false) :
+                new();
+
+            if (request.Params?.Cursor is null && resourceCollection is not null)
+            {
+                result.Resources.AddRange(resourceCollection.Select(t => t.ProtocolResource));
+            }
 
-        listResourcesHandler ??= (static (_, _) => Task.FromResult(new ListResourcesResult()));
+            return result;
+        };
+
+        var isMissingListResourceHandlers = originalListResourcesHandler is not { } && listResourceTemplatesHandler is not { };
+        if (resourceCollection is not { IsEmpty: false } && (isMissingListResourceHandlers || readResourceHandler is not { }))
+        {
+            throw new McpException("Resources capability was enabled, but ListResources, ListResourceTemplates, and/or ReadResource handlers were not specified.");
+        }
```

Comment:
I don't understand this logic. Why is "is missing" true if both originalListResourcesHandler and listResourceTemplatesHandler are non-null?